ouroboros-consensus-test/test-storage/Test/Ouroboros/Storage/LedgerDB/DbChangelog.hs

```diff
@@ -0,0 +1,427 @@
+
+prop_flushingSplitsImmutableAndVolatile :: DbChangelogTestSetup -> Property
+prop_flushingSplitsImmutableAndVolatile setup =
+  property $ AS.length (changelogVolatileStates toFlush) == 0 &&
+             AS.length (changelogImmutableStates toKeep) == 0 &&
+             diffSeqDblog == diffSeqJoined
+  where
+    dblog = resultingDbChangelog setup
+    (toFlush, toKeep) = flushDbChangelog DbChangelogFlushAllImmutable dblog
+    (ApplySeqDiffMK diffSeqToFlush) = unTestTables $ changelogDiffs toFlush
+    (ApplySeqDiffMK diffSeqToKeep) = unTestTables $ changelogDiffs toKeep
+    diffSeqJoined = unsafeJoinSeqUtxoDiffs diffSeqToFlush diffSeqToKeep
+    (ApplySeqDiffMK diffSeqDblog) = unTestTables $ changelogDiffs dblog
+
+prop_prefixBackToAnchorKeepsInvariants :: DbChangelogTestSetup -> Property
+prop_prefixBackToAnchorKeepsInvariants setup = property $ checkInvariants dblog
+  where
+    dblog = prefixBackToAnchorDbChangelog $ resultingDbChangelog setup
+
+prop_flushDbChangelogKeepsInvariants :: DbChangelogTestSetup -> Property
+prop_flushDbChangelogKeepsInvariants setup =
+  property $ checkInvariants toFlush && checkInvariants toKeep
+  where
+    (toFlush, toKeep) = flushDbChangelog DbChangelogFlushAllImmutable $
+      resultingDbChangelog setup
+
+prop_rollbackDbChangelogKeepsInvariants ::
+  DbChangelogTestSetupWithRollbacks -> Property
+prop_rollbackDbChangelogKeepsInvariants setup = property $ checkInvariants dblog
+  where
+    n = rollbacks setup
+    dblog = rollbackDbChangelog n (resultingDbChangelog $ testSetup setup)
+
+prop_prefixBackToAnchorIsRollingBackVolatileStates :: DbChangelogTestSetup -> Property
+prop_prefixBackToAnchorIsRollingBackVolatileStates setup =
+  property $ rolledBack == toAnchor
+  where
+    dblog = resultingDbChangelog setup
+    n = AS.length (changelogVolatileStates dblog)
+    rolledBack = rollbackDbChangelog n dblog
+    toAnchor = prefixBackToAnchorDbChangelog dblog
+
+prop_rollBackToVolatileTipIsNoop ::
+  Positive Int -> DbChangelogTestSetup -> Property
+prop_rollBackToVolatileTipIsNoop (Positive n) setup = property $ Just dblog == dblog'
+  where
+    dblog = resultingDbChangelog setup
+    pt = getTip $ unDbChangelogState $ AS.headAnchor $ changelogVolatileStates dblog
+    dblog' = prefixDbChangelog pt $ nExtensions n dblog
+
+unsafeJoinSeqUtxoDiffs :: Ord k => SeqUtxoDiff k v -> SeqUtxoDiff k v -> SeqUtxoDiff k v
+unsafeJoinSeqUtxoDiffs (SeqUtxoDiff ft1) (SeqUtxoDiff ft2) =
+  SeqUtxoDiff (ft1 FT.>< ft2)
+
+
+{-------------------------------------------------------------------------------
+  Generators
+-------------------------------------------------------------------------------}
+
+pointAtSlot :: WithOrigin SlotNo -> Point (TestLedger EmptyMK)
+pointAtSlot = Point.withOrigin GenesisPoint (\slotNo -> Point $ At $ Point.Block slotNo H)
+
+type Key = String
+
+data GenOperationsState = GenOperationsState {
+    osSlotNo            :: !(WithOrigin SlotNo)
```

Comment:
Maybe we should comment what these fields are. 

Also, do we need these fields to be strict?